### ISSUE TITLE
Conditionally Mapped Type Leads to Confusing ts(2345) Error

### ISSUE THREAD

**Issue by SlurpTheo** [microsoft/TypeScript#40550]
**TypeScript Version:**
3.9.5
4.1.0-dev.20200914

**Search Terms:**
2345
conditional

**Code**

```ts
declare type GetPropertyNamesOfType<T, RestrictToType> = {
	[K in keyof T]: T[K] extends RestrictToType ? K : never;
}[keyof T];
```
You can find this gnarly type here:
https://github.com/microsoft/TypeScript/blob/94d4023043dd032a8c96a59677eb63526b6b24b6/tests/cases/conformance/types/conditional/conditionalTypes2.ts#L227

The following consuming/using code _correctly_ generates `ts(2345)` errors for the last two lines:
```ts
interface A {
	req: string;
	num: number;
	str: string;
}
interface B {
	req: string;
	num?: number;
	str?: string;
}

declare function callA(name: GetPropertyNamesOfType<A, string>): void;
declare function callB(name: GetPropertyNamesOfType<Required<B>, string>): void;

const prop = "someKey";
callA(prop); // ts(2345)
callB(prop); // ts(2345)
```

**Expected behavior:**
In `typescript@3.8` (and a few earlier versions I tried) the error message for both lines:
```
Argument of type '"someKey"' is not assignable to parameter of type '"req" | "str"'.
```

**Actual behavior:**
In `typescript@^3.9` this is reported for both lines:
```
Argument of type '"someKey"' is not assignable to parameter of type 'GetPropertyNamesOfType<A, string>'.
```

That's correct (nice even!) for the `callA()` line, but it's confusing for the `callB()` line.

(This reproduction is a single file that can run into this problem... it's not how I ran into it. I had a `ts.Program` with many 10s of `*.ts` files in the same compilation and the message was referring to some type that made no sense in the source file that was giving me the error.)

**Playground Link:**
https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAXAngBwQcRIgCjA9ujCgHJQC2IAzgPIBmAKmiADwMA08AStYjAJZhEDfE3QA+eAF54AbwCwAKACQAbQDS8fgDt4AaxDJ8deAwC6ALlMaz8EAA9EIbcCrdeAoSLEIA-PE0rbRAANxAYAG4lAF9VAyMTcyjFJR0nGDooMAQAQTklZTgARysqPh0Ac2TlbQBXciD6gCNw6rKYUvLtKpjU7XTM7PgAIXyVYs6Bbuq68l9G8hbIgvb5+HbK5OilJVBIWAQ6Wu0hfnxdMCgICByACm0KECssXAIiUkfaRmYWHM4N7riACUVhC+H4wGSe2gcHgRxOiDOFyuEGG90ez2weEI4Q+lC+PhYPCKtX4cGALGG4n+XQqwNB4MhO0UYHOZXgqDe0ngACIqPhKOpDDzkpdrndOYQgRF4AB6WVIKi3ABMAGYACwAViBSjFqNuktQ0rlCsQSrVWp1iiAA

**Related Issues:**
I found reference to the gnarly type above in #32608 & #33568.
It looks to be the same "monster" as you see in #29505 too.

**Comment by RyanCavanaugh** [comment]
I'm not sure that this is better or worse. If the union had hundreds of elements, as is common in many JSX scenarios, then it'd be much worse to just see a big list of string literals without any context as to where they came from. If the type alias name is descriptive, then a human reading it should be able to understand it from a higher-order perspective.

Tagging Needs Proposal if someone wants to figure out some motivating set of use cases for each behavior and determine what the right way to figure out the "correct" display is.

**Comment by SlurpTheo** [comment]
You know, I was picturing a response much like this from you as I was writing this issue 😺
I hear ya.
I mainly submitted to make sure this was a deliberate, desired change beginning with `typescript@3.9`.

At some point union literals _do_ stop listing all constituents with the `| ... 4 more ... |` portion of the below error message.
```ts
declare function function2(p?: keyof tLotsAProps2): void;
type tLotsAProps2 = {
	p10: string;
	p11: string;
	p12: string;
	p13: string;
	p14: string;
	p15: string;
	p16: string;
	p17: string;
	p18: string;
	p19: string;
	p20: string;
	p21: string;
	p22: string;
	p23: string;
	p24: string;
	p25: string;
	p26: string;
	p27: string;
	p28: string;
	p29: string;
	p30: string;
	p31: string;
	p32: string;
	p33: string;
	p34: string;
	p35: string;
	p36: string;
};
const prop2 = "someKey";
function2(prop2); // ts(2345)
```
```
Argument of type '"someKey"' is not assignable to parameter of type '"p10" | "p11" | "p12" | "p13" | "p14" | "p15" | "p16" | "p17" | "p18" | "p19" | "p20" | "p21" | "p22" | "p23" | "p24" | "p25" | "p26" | "p27" | "p28" | "p29" | "p30" | "p31" | "p32" | ... 4 more ... | undefined'.
```
Now, as you say, without context that might not be too helpful either... there will be no winning I guess.

I could imagine, with inspiration from the current conditional/mapped types experience, this error instead being:
```
Argument of type '"someKey"' is not assignable to parameter of type 'keyof tLotsAProps2 | undefined'.
```

**Comment by SlurpTheo** [comment]
I tweaked the main text just a bit.
The behaviour change was between `typescript@3.9.5` & `typescript@3.9.6`.
